src/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs

```diff
@@ -0,0 +1,841 @@
+
+        private readonly IBufferWriter<byte> _output;
+        private int _buffered;
+        private Span<byte> _buffer;
+
+        /// <summary>
+        /// Returns the total amount of bytes written by the <see cref="Utf8JsonWriter"/> so far
+        /// for the current instance of the <see cref="Utf8JsonWriter"/>.
+        /// This includes data that has been written beyond what has already been committed.
+        /// </summary>
+        public long BytesWritten
+        {
+            get
+            {
+                Debug.Assert(BytesCommitted <= long.MaxValue - _buffered);
+                return BytesCommitted + _buffered;
+            }
+        }
+
+        /// <summary>
+        /// Returns the total amount of bytes committed to the output by the <see cref="Utf8JsonWriter"/> so far
+        /// for the current instance of the <see cref="Utf8JsonWriter"/>.
+        /// This is how much the IBufferWriter has advanced.
+        /// </summary>
+        public long BytesCommitted { get; private set; }
+
+        private bool _inObject;
+        private bool _isNotPrimitive;
+        private JsonTokenType _tokenType;
+        private readonly JsonWriterOptions _writerOptions;
+        private BitStack _bitStack;
+
+        // The highest order bit of _currentDepth is used to discern whether we are writing the first item in a list or not.
+        // if (_currentDepth >> 31) == 1, add a list separator before writing the item
+        // else, no list separator is needed since we are writing the first item.
+        private int _currentDepth;
+
+        private int Indentation => CurrentDepth * SpacesPerIndent;
+
+        /// <summary>
+        /// Tracks the recursive depth of the nested objects / arrays within the JSON text
+        /// written so far. This provides the depth of the current token.
+        /// </summary>
+        public int CurrentDepth => _currentDepth & JsonConstants.RemoveFlagsBitMask;
+
+        /// <summary>
+        /// Returns the current snapshot of the <see cref="Utf8JsonWriter"/> state which must
+        /// be captured by the caller and passed back in to the <see cref="Utf8JsonWriter"/> ctor with more data.
+        /// Unlike the <see cref="Utf8JsonWriter"/>, which is a ref struct, the state can survive
+        /// across async/await boundaries and hence this type is required to provide support for reading
+        /// in more data asynchronously before continuing with a new instance of the <see cref="Utf8JsonWriter"/>.
+        /// </summary>
+        public JsonWriterState CurrentState
+        {
+            get
+            {
+                // Getting the state for creating a new Utf8JsonWriter without first committing the data that has been written
+                // would result in an inconsistent state. Therefore, calling Flush before getting the current state.
+                if (_buffered != 0)
+                {
+                    Flush();
```

Comment:
Should we Flush here on the user's behalf or change this to a method and throw an exception?

Reply:
@KrzysztofCwalina earlier was talking about Flush vs FlushAsync and such.  If the required usage is to call Flush(Async) before being able to snap the state, it seems like a method is the way to go.  It also stops a heisen-bug of the debugger ends up calling Flush for you every time you hit F10.

Reply:
We end up losing parity with the `Utf8JsonReader.CurrentState` property if we change it to a method, which is why I was flushing for the caller. Are we fine with it diverging here?

Reply:
There's still time to change the reader version to a method; too.  But the property having a side effect of calling Flush seems to me to violate property rules; so this one needs to change.

Properties behaving like properties is more important than "property on the reader => property on the writer".